ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -34,6 +36,91 @@ module Winbind = struct
     debug "init_service To be implemented in CP-36085"
 end
 
+let net_cmd = "/usr/bin/net"
+
+let from_config ~name ~err_msg ~config_params =
+  match List.assoc_opt name config_params with
+  | Some v ->
+      v
+  | _ ->
+      raise (Auth_service_error (E_GENERIC, err_msg))
+
+let assert_hostname_valid () =
+  let hostname =
+    (fun __context ->
+      Helpers.get_localhost ~__context |> fun host ->
+      Db.Host.get_hostname ~__context ~self:host)
+    |> Server_helpers.exec_with_new_task "retrieving hostname"
+  in
+  let all_numbers hostname =
+    String.fold_left (fun b ch -> b && ch >= '0' && ch <= '9') true hostname
+  in
+  match all_numbers hostname with
+  | true ->
+      raise
+        (Auth_service_error
+           ( E_GENERIC
+           , Printf.sprintf "hostname '%s' cannot contain only digits." hostname
+           ))
+  | false ->
+      ()
+
+let assert_domain_equal_service_name ~service_name ~config_params =
+  (* For legeacy support, if domain exist in config_params, it must be equal to service_name *)
+  let domain_key = "domain" in
+  if List.mem_assoc domain_key config_params then
+    let domain = List.assoc domain_key config_params in
+    if domain <> service_name then
+      raise
+        (Auth_service_error
+           (E_GENERIC, "if present, config:domain must match service-name."))
+
+let extract_ou_config ~config_params =
+  try
+    let ou = from_config ~name:"ou" ~err_msg:"" ~config_params in
+    ([("ou", ou)], [Printf.sprintf "createcomputer=%s" ou])
+  with Auth_service_error _ -> ([], [])
+
+let query_domain_workgroup domain =
+  let key = "Pre-Win2k Domain" in
+  let lines =
+    Helpers.call_script ~log_output:On_failure net_cmd
```

Comment:
This will raise an exception on failure, I see you intend to handle failure below, in which case you need to catch `Forkhelpers.Spawn_internal_error (stderr, stdout, status) as e`, see `helpers.ml`